vault/resource_ad_secret_library.go

```diff
@@ -0,0 +1,232 @@
+				return strings.Trim(v.(string), "/")
+			},
+		},
+		"name": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: `The name of the set of service accounts.`,
+			ForceNew:    true,
+		},
+		"service_account_names": {
+			Type:        schema.TypeList,
+			Elem:        &schema.Schema{Type: schema.TypeString},
+			Required:    true,
+			Description: "The names of all the service accounts that can be checked out from this set. These service accounts must already exist in Active Directory.",
+			ForceNew:    true,
+		},
+		"max_ttl": {
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Computed:    true,
+			Description: `The maximum amount of time a check-out last with renewal before Vault automatically checks it back in. Defaults to 24 hours. Setting it to zero reflects an unlimited lending period.`,
+		},
+		"ttl": {
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Computed:    true,
+			Description: `The maximum amount of time a single check-out lasts before Vault automatically checks it back in. Defaults to 24 hours. Setting it to zero reflects an unlimited lending period.`,
+		},
+		"disable_check_in_enforcement": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Description: `Disable enforcing that service accounts must be checked in by the entity or client token that checked them out.`,
+		},
+	}
+	return &schema.Resource{
+		Create: createLibraryResource,
+		Update: updateLibraryResource,
+		Read:   readLibraryResource,
+		Delete: deleteLibraryResource,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: fields,
+	}
+}
+
+func createLibraryResource(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	backend := d.Get("backend").(string)
+	set := d.Get("name").(string)
+	setPath := fmt.Sprintf("%s/library/%s", backend, set)
+
+	log.Printf("[DEBUG] Creating %q", setPath)
+
+	data := map[string]interface{}{}
+	if v, ok := d.GetOkExists("disable_check_in_enforcement"); ok {
+		data["disable_check_in_enforcement"] = v
+	}
+
+	if v, ok := d.GetOkExists("service_account_names"); ok {
+		data["service_account_names"] = v
+	}
+
+	if v, ok := d.GetOkExists("max_ttl"); ok {
+		data["max_ttl"] = v
+	}
+
+	if v, ok := d.GetOkExists("ttl"); ok {
+		data["ttl"] = v
+	}
+
+	log.Printf("[DEBUG] Writing %q", setPath)
+	if _, err := client.Logical().Write(setPath, data); err != nil {
+		return fmt.Errorf("error writing %q: %s", setPath, err)
+	}
+	d.SetId(setPath)
+	log.Printf("[DEBUG] Wrote %q", setPath)
+	return readLibraryResource(d, meta)
+}
+
+func readLibraryResource(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	setPath := d.Id()
+	log.Printf("[DEBUG] Reading %q", setPath)
+
+	setName, err := adSecretBackendSetNameFromPath(setPath)
+	if err != nil {
+		log.Printf("[WARN] Removing AD library %q because its ID is invalid", setPath)
+		d.SetId("")
```

Comment:
I don't feel like we should remove from state here. My reasoning is that we're doing a name regex match and if it fails, we remove the thing from state. It's possible in the future the supported names / regex patterns in Vault could change, and we could inappropriately remove things from state because Terraform has an out of date regex. I'm not sure why we need to ensure our ID conforms to a regex at time of READ either 🤔 

For removing from state, we should typically reserve that for when Vault returns a 404 for that resource.

Reply:
Done